### PR TITLE
Add PR sheriff automation

### PR DESCRIPTION
**Pull request by pbakaus**
This PR adds a PR sheriff GitHub Action for pull request triage.

---

> [!NOTE]
> **Medium Risk**
> The workflow can close PRs and rewrite labels on every open PR when `--apply` runs; logic is heavily tested but mistakes would affect all contributors’ PRs.
> 
> **Overview**
> Adds automated **PR sheriff** triage: a scheduled GitHub Action (daily, plus manual `workflow_dispatch` with optional dry-run) runs `scripts/github/sheriff.mjs` with write access to issues and pull requests.
> 
> The script loads open PRs via GraphQL/`gh`, ensures a defined label set, and for each PR computes a plan: sync state labels (`waiting on contributor`, `needs maintainer review`, `ready to merge`) and blocker labels (CI, review threads, merge conflicts), detect contributor vs maintainer responsibility (draft, failing checks, conflicts, `CHANGES_REQUESTED`, unresolved threads, maintainer `/sheriff wait`, timed `waiting on contributor` labels), post **stale warnings** after 7 days and **auto-close** after 14 when still blocked (with exemptions for `do not close` / `security` and default skip auto-close for listed regular contributors). Apply mode mutates labels, comments, and can close PRs.
> 
> **Tests:** `tests/github-sheriff.test.mjs` covers evaluation edge cases; `scripts/test-suites.mjs` wires it into the core suite.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ab23b00a7fbd521d74d607da74ad5a6c16441eb8. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: scripts/github/sheriff.mjs [hunk 1]
```diff
@@ -0,0 +1,629 @@
...
+        type: event.__typename,
+        label: event.label.name,
+        actorLogin: event.actor?.login || '',
+        createdAt: event.createdAt,
+      })),
+  };
+}
+
+export function staleWarningComment(pr, { daysOpen, closeDays }) {
+  const closeDate = formatDate(addDays(toDate(pr.createdAt), closeDays));
```
[the remark this answers is not in the document]
Comment (pbakaus) [reply scripts/github/sheriff.mjs]: Fixed in 933881c: the stale warning close date now clamps to at least the next day for already-overdue PRs, with regression coverage.

File: .github/workflows/sheriff.yml [hunk 2]
```diff
@@ -0,0 +1,50 @@
+name: PR Sheriff
```
[the remark this answers is not in the document]
Comment (pbakaus) [reply .github/workflows/sheriff.yml:1]: Fixed: updated the PR title and opening description to describe PR sheriff automation and pull request triage instead of issue triage.

File: scripts/github/sheriff.mjs [hunk 3]
```diff
@@ -0,0 +1,629 @@
...
+  const closeAlreadyPosted = hasMarker(pr.comments, CLOSE_MARKER);
+  const exemptFromClose = [...labels].some((label) => exemptLabels.has(label));
+  const regularContributor = regularContributors.has(author);
+  const shouldWarn = staleEligible && !warningAlreadyPosted;
+  const shouldClose = contributorActionRequired
+    && daysOpen >= closeDays
+    && warningAlreadyPosted
+    && !closeAlreadyPosted
+    && !exemptFromClose
+    && (autoCloseRegulars || !regularContributor);
```
[the remark this answers is not in the document]
Comment (pbakaus) [reply scripts/github/sheriff.mjs:214]: Fixed in 933881c: the stale label now acts as durable proof that the sheriff warning already happened even if older marker comments fell out of the fetched comment window. Added regression coverage.

File: scripts/github/sheriff.mjs [hunk 4]
```diff
@@ -0,0 +1,629 @@
...
+function latestMaintainerAsk(pr, maintainers) {
+  return latestDate([
+    ...(pr.comments || [])
+      .filter((comment) => maintainers.has(normalizeLogin(comment.authorLogin)))
+      .map((comment) => comment.createdAt),
+    ...(pr.reviews || [])
+      .filter((review) => maintainers.has(normalizeLogin(review.authorLogin)))
+      .filter((review) => REVIEW_MAINTAINER_STATES.has(review.state))
+      .map((review) => review.submittedAt),
+  ]);
```
[the remark this answers is not in the document]
Comment (pbakaus) [reply scripts/github/sheriff.mjs]: Fixed in 933881c: maintainer issue comments and COMMENTED reviews now count only when they look actionable, so LGTM or other non-action maintainer chatter will not stale the PR. Added regression coverage.

File: scripts/github/sheriff.mjs [hunk 5]
```diff
@@ -0,0 +1,629 @@
...
+  }
+
+  if (latestMaintainerAskAt && !isAfter(latestContributorAt, latestMaintainerAskAt)) {
+    blockers.push({ kind: 'maintainer-ask', at: latestMaintainerAskAt });
+  }
+
+  const waitingLabelAt = latestLabelEventAt(pr.labelEvents || [], 'waiting on contributor', 'LabeledEvent');
+  if (labels.has('waiting on contributor') && waitingLabelAt && !isAfter(latestContributorAt, waitingLabelAt)) {
+    blockers.push({ kind: 'manual-waiting', at: waitingLabelAt });
+  }
```
[the remark this answers is not in the document]
Comment (pbakaus) [reply scripts/github/sheriff.mjs:179]: Fixed in 933881c: a current waiting label is preserved when its LabeledEvent is missing from the fetched timeline, while still clearing after newer contributor action when we can infer it. Added regression coverage.

File: .github/workflows/sheriff.yml [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    inputs:
+      dry_run:
+        description: "Print planned changes without mutating GitHub"
+        type: boolean
+        default: false
+
+permissions:
+  contents: read
+  issues: write
+  pull-requests: write
```
[the remark this answers is not in the document]
Comment (pbakaus) [reply .github/workflows/sheriff.yml:20]: Fixed in 933881c: the workflow now grants actions:read and checks:read so statusCheckRollup remains available under explicit token permissions.

File: .github/workflows/sheriff.yml [hunk 7]
```diff
@@ -0,0 +1,50 @@
+name: PR Sheriff
+
+on:
+  schedule:
+    # Daily UTC morning pass. The script uses an aggressive 7/14 day window:
+    # warn contributor-blocked PRs after 7 days open, close after 14.
+    - cron: "17 15 * * *"
```
[the remark this answers is not in the document]
Comment (pbakaus) [reply .github/workflows/sheriff.yml]: Fixed in 933881c: the workflow comment now says UTC afternoon, matching the 15:17 UTC cron schedule.

File: scripts/github/sheriff.mjs [hunk 8]
```diff
@@ -0,0 +1,646 @@
...
+      };
+    })
+    .filter((thread) => thread.latestCommentAt)
+    .filter((thread) => thread.latestCommentAuthor !== author);
+}
+
+function applyPlan(repo, plan, options) {
+  for (const label of plan.labelsToRemove) removeLabel(repo, plan.number, label, options);
+  if (plan.labelsToAdd.length > 0) addLabels(repo, plan.number, plan.labelsToAdd, options);
+  if (plan.shouldWarn) postComment(repo, plan.number, plan.warningComment, options);
```
[the remark this answers is not in the document]
Comment (pbakaus) [reply scripts/github/sheriff.mjs]: Fixed in 2582a5c: warning comments are now posted before label mutations, so a failed warning comment cannot leave behind a stale label that later suppresses the warning path.

File: scripts/github/sheriff.mjs [hunk 9]
```diff
@@ -0,0 +1,646 @@
...
+
+  if (latestMaintainerWaitAt && !isAfter(latestContributorAt, latestMaintainerWaitAt)) {
+    blockers.push({ kind: 'sheriff-wait', at: latestMaintainerWaitAt });
+  }
+
+  const waitingLabelAt = latestLabelEventAt(pr.labelEvents || [], 'waiting on contributor', 'LabeledEvent');
+  const waitingLabelReferenceAt = waitingLabelAt || pr.updatedAt;
+  if (labels.has('waiting on contributor') && !isAfter(latestContributorAt, waitingLabelReferenceAt)) {
+    blockers.push({ kind: 'manual-waiting', at: waitingLabelReferenceAt });
+  }
```
[the remark this answers is not in the document]
Comment (pbakaus) [reply scripts/github/sheriff.mjs:179]: Fixed in 2582a5c: the sheriff now hydrates missing waiting-label timestamps from issue events and no longer falls back to pr.updatedAt, with regression coverage for the no-timestamp case.

File: scripts/github/sheriff.mjs [hunk 10]
```diff
@@ -0,0 +1,705 @@
...
+  if (FAILING_STATUS_STATES.has(pr.statusState)) {
+    blockers.push({ kind: 'ci', label: 'blocked: ci', at: pr.latestCommitAt || pr.updatedAt });
+  }
+
+  if (pr.mergeable === 'CONFLICTING') {
+    blockers.push({ kind: 'merge-conflicts', label: 'blocked: merge conflicts', at: pr.updatedAt });
+  }
+
+  if (latestBlockingReviewAt && !isAfter(latestContributorAt, latestBlockingReviewAt)) {
+    blockers.push({ kind: 'changes-requested', label: 'blocked: review threads', at: latestBlockingReviewAt });
```
[the remark this answers is not in the document]
Comment (pbakaus) [reply scripts/github/sheriff.mjs]: Fixed in 96becc4: changes-requested review timestamps are now only treated as active blockers when GitHub’s current reviewDecision is still CHANGES_REQUESTED. I added regression coverage for both the cleared-review path and the still-current blocker path.

File: scripts/github/sheriff.mjs [hunk 11]
```diff
@@ -0,0 +1,707 @@
...
+
+  const labels = new Set(pr.labels || []);
+  const author = normalizeLogin(pr.authorLogin || pr.author?.login || '');
+  const daysOpen = Math.floor((now.getTime() - toDate(pr.createdAt).getTime()) / DAY_MS);
+  const latestContributorAt = latestDate([
+    pr.latestCommitAt,
+    ...commentsBy(pr.comments, author).map((comment) => comment.createdAt),
+    ...reviewsBy(pr.reviews, author).map((review) => review.submittedAt),
+    ...reviewThreadCommentsBy(pr.reviewThreads, author).map((comment) => comment.createdAt),
+  ]);
```
[the remark this answers is not in the document]
Comment (pbakaus) [reply scripts/github/sheriff.mjs:138]: Fixed in ab23b00: latest head commits now only count as contributor activity when the commit author or committer matches the PR author. Non-author branch updates keep /sheriff wait and manual waiting labels active, with regression coverage for that case.
